stdlib/regex.gr

```diff
@@ -0,0 +1,2983 @@
+      None => None
+    }
+  }
+  let getAllMatchGroupPositions = () => {
+    let ret = Array.make(Array.length(state) + 1, None)
+    ret[0] = Some((start, end))
+    for (let mut i = 0; i < Array.length(state); i += 1) {
+      ret[i + 1] = state[i]
+    }
+    ret
+  }
+  let getAllMatchGroups = () => {
+    Array.map(o => match(o) {
+      None => None,
+      Some((start, end)) => Some(String.slice(start, end, origString))
+    }, getAllMatchGroupPositions())
+  }
+  {
+    group: getMatchGroup,
+    groupPosition: getMatchGroupPosition,
+    numGroups: Array.length(state) + 1,
+    allGroupPositions: getAllMatchGroupPositions,
+    allGroups: getAllMatchGroups
+  }
+}
+
+// Helpers for user-facing match functionality
+
+let fastDriveRegexIsMatch = (rx, string, startOffset, endOffset) => {
+  let state = if (rx.reReferences) Array.make(rx.reNumGroups, None) else Array.make(0, None)
+  let toWrap = if (startOffset == 0 && endOffset == String.length(string)) string else String.slice(startOffset, endOffset, string)
+  let buf = makeMatchBuffer(toWrap)
+  Option.isSome(searchMatch(rx, buf, 0, 0, Array.length(buf.matchInputExploded), state))
+}
+
+let rec fastDriveRegexMatchAll = (rx, string, startOffset, endOffset) => {
+  if (startOffset >= endOffset) {
+    []
+  } else {
+    let state = Array.make(rx.reNumGroups, None)
+    let toWrap = if (startOffset == 0 && endOffset == String.length(string)) string else String.slice(startOffset, endOffset, string)
+    let buf = makeMatchBuffer(toWrap)
+    match(searchMatch(rx, buf, 0, 0, Array.length(buf.matchInputExploded), state)) {
+      None => [],
+      Some((startPos, endPos)) => [makeMatchResult(string, startPos + startOffset, endPos + startOffset, Array.map(elt => {
+        match(elt) {
+          None => None,
+          Some((start, end)) => Some((start + startOffset, end + startOffset))
+        }
+      }, state)), ...fastDriveRegexMatchAll(rx, string, startPos + startOffset + 1, endOffset)],
+    }
+  }
+}
+
+let fastDriveRegexMatch = (rx, string, startOffset, endOffset) => {
+  let state = Array.make(rx.reNumGroups, None)
+  let toWrap = if (startOffset == 0 && endOffset == String.length(string)) string else String.slice(startOffset, endOffset, string)
+  let buf = makeMatchBuffer(toWrap)
+  match(searchMatch(rx, buf, 0, 0, Array.length(buf.matchInputExploded), state)) {
+    None => None,
+    Some((startPos, endPos)) => {
+      Some(makeMatchResult(string, startPos + startOffset, endPos + startOffset, Array.map(elt => {
+        match(elt) {
+          None => None,
+          Some((start, end)) => Some((start + startOffset, end + startOffset))
+        }
+      }, state)))
+    }
+  }
+}
+
+/**
+ * Returns true if the given regular expression has a match in the given string
+ * @param rx: RegularExpression - The regular expression to search for
+ * @param string: String - The string to search within
+ * @returns Whether `rx` matches `string`
```

Comment:
```suggestion
 * Determines if the given regular expression has a match in the given string.
 * @param rx: The regular expression to search for
 * @param string: The string to search within
 * @returns `true` if the RegExp matches the string, otherwise `false`
```